File: backend/question_scraper.py
import requests
from bs4 import BeautifulSoup

REQUEST_HEADERS = {
    "User-Agent": "AI-Interview-Simulator/1.0",
}


def fetch_questions_from_web(skill):

    url = f"https://www.geeksforgeeks.org/{skill.replace(' ', '-')}-interview-questions/"

    questions = []

    try:
        response = requests.get(url, headers=REQUEST_HEADERS, timeout=10)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, "html.parser")

        for li in soup.find_all("li"):
            text = li.get_text(" ", strip=True)

            if "?" in text:
                questions.append(text.strip())

    except requests.RequestException:
        return []

    return questions[:5]

File: backend/src/services/questionScraper.js
const https = require("https");
const http = require("http");
const { JSDOM } = require("jsdom");

const REQUEST_HEADERS = {
  "User-Agent": "AI-Interview-Simulator/1.0",
};

/**
 * Fetch interview questions for a skill from GeeksForGeeks.
 * Returns up to 5 questions.
 * @param {string} skill
 * @returns {Promise<string[]>}
 */
async function fetchQuestionsFromWeb(skill) {
  const slug = skill.replace(/\s+/g, "-");
  const url = `https://www.geeksforgeeks.org/${slug}-interview-questions/`;

  return new Promise((resolve) => {
    const mod = url.startsWith("https") ? https : http;

    const req = mod.get(url, { headers: REQUEST_HEADERS, timeout: 10000 }, (res) => {
      let body = "";
      res.on("data", (chunk) => (body += chunk));
      res.on("end", () => {
        try {
          const dom = new JSDOM(body);
          const items = dom.window.document.querySelectorAll("li");
          const questions = [];

          items.forEach((li) => {
            const text = (li.textContent || "").trim();
            if (text.includes("?")) {
              questions.push(text);
            }
          });

          resolve(questions.slice(0, 5));
        } catch {
          resolve([]);
        }
      });
    });

    req.on("error", () => resolve([]));
    req.on("timeout", () => {
      req.destroy();
      resolve([]);
    });
  });
}

module.exports = { fetchQuestionsFromWeb };
